File: js/logic/GraphState.js
/**
 * @fileoverview Manages the internal state of the argumentation graph by communicating with the backend.
 */

export class GraphState {
  constructor(data) {
    this.apiUrl = 'http://localhost:8000';
    this.colors = data.colors || {};
    this.cats = data.cats || {};
    this.globalScoresConfig = data.globalScores || [];
    this.method = 'h-categorizer';
    this.nodes = data.nodes || [];
    this.links = [];
    this.rebuildLinks();

    this.onUpdate = () => {};
    this.onStructureChange = () => {};
    this.onVisualChange = () => {};
    this.onNodeClick = () => {};
    this.onSelectionChange = () => {};
    
    this.selectedNodeId = null;
  }

  async recalculate() {
    const res = await fetch(`${this.apiUrl}/graph`);
    const data = await res.json();
    this.updateLocalData(data);
    this.onVisualChange();
  }

  updateLocalData(data) {
    this.nodes = data.nodes;
    this.rebuildLinks();
    // Signal sidebar update with new category averages
    const categoryScores = this.calculateCategoryAverages();
    this.onUpdate(categoryScores);
  }

  calculateCategoryAverages() {
    const averages = {};
    const categories = Object.keys(this.cats).filter(c => c !== 'action' && c !== 'state');
    categories.forEach(cat => {
      const nodes = this.nodes.filter(n => n.cat === cat && !n.inactive);
      if (nodes.length > 0) {
        averages[cat] = nodes.reduce((sum, n) => sum + (n.score || 0), 0) / nodes.length;
      } else {
        averages[cat] = 0;
      }
    });
    return averages;
  }

  rebuildLinks() {
    this.links = [];
    this.nodes.forEach(n => {
      (n.attacks || []).forEach(targetId => {
        this.links.push({ source: n.id, target: targetId, type: 'attack' });
      });
      (n.supports || []).forEach(targetId => {
        this.links.push({ source: n.id, target: targetId, type: 'support' });
      });
    });
  }

  async addNode(nodeObj) {
    const res = await fetch(`${this.apiUrl}/node`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(nodeObj)
    });
    const data = await res.json();
    this.updateLocalData(data);
    this.onStructureChange();
  }

  async updateNode(id, newData) {
    const structural = newData.id !== undefined || newData.cat !== undefined || newData.attacks !== undefined || newData.supports !== undefined;

    const res = await fetch(`${this.apiUrl}/node/${id}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(newData)
    });
    const data = await res.json();
    this.updateLocalData(data);
    
    if (structural) {
      this.onStructureChange();
    } else {
      this.onVisualChange();
    }
  }

  async removeNode(id) {
    const res = await fetch(`${this.apiUrl}/node/${id}`, {
      method: 'DELETE'
    });
    const data = await res.json();
    this.updateLocalData(data);
    this.onStructureChange();
  }

  async toggleNodeActive(id) {
    const node = this.getNode(id);
    if (node && node.cat !== 'state') {
      await this.updateNode(id, { inactive: !node.inactive });
    }
  }

getNode(id) {
  return this.nodes.find(n => n.id === id) || null;
}

/**
 * Exports the current state to JSON.
 */
exportJSON() {
    const data = {
      colors: this.colors,
      cats: this.cats,
      globalScores: this.globalScoresConfig,
      nodes: this.nodes.map(n => ({
        id: n.id,
        label: n.label,
        cat: n.cat,
        desc: n.desc,
        attacks: [...(n.attacks || [])],
        supports: [...(n.supports || [])],
        value: n.value,
        inactive: n.inactive
      }))
    };
    
    const blob = new Blob([JSON.stringify(data, null, 2)], { type: 'application/json' });
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = 'argumentation_graph.json';
    a.click();
    URL.revokeObjectURL(url);
  }
}

File: backend/graph/graph_state.py
from typing import List, Dict, Any, Optional
from .semantics import calculate_h_categorizer, calculate_max_based

class Node:
    def __init__(self, data: Dict[str, Any]):
        self.id = data.get("id")
        self.label = data.get("label")
        self.cat = data.get("cat")
        self.desc = data.get("desc")
        self.value = data.get("value", 0.5)
        self.score = data.get("score", 0.0)
        self.attacks = data.get("attacks", [])
        self.supports = data.get("supports", [])
        self.inactive = data.get("inactive", False)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "label": self.label,
            "cat": self.cat,
            "desc": self.desc,
            "value": self.value,
            "score": self.score,
            "attacks": self.attacks,
            "supports": self.supports,
            "inactive": self.inactive
        }

class GraphState:
    def __init__(self, data: Dict[str, Any]):
        self.colors = data.get("colors", {})
        self.cats = data.get("cats", {})
        self.global_scores_config = data.get("globalScores", [])
        self.nodes = [Node(n) for n in data.get("nodes", [])]
        self.method = "h-categorizer"
        self.category_weights = {}

    def recalculate(self):
        if self.method == "max-based":
            scores = calculate_max_based(self.nodes, self.category_weights)
        else:
            scores = calculate_h_categorizer(self.nodes, self.category_weights)
            
        for node in self.nodes:
            node.score = scores.get(node.id, 0.0)

    def update_node(self, node_id: str, data: Dict[str, Any]):
        for node in self.nodes:
            if node.id == node_id:
                for k, v in data.items():
                    setattr(node, k, v)
                break
        self.recalculate()

    def add_node(self, data: Dict[str, Any]):
        self.nodes.append(Node(data))
        self.recalculate()

    def remove_node(self, node_id: str):
        self.nodes = [n for n in self.nodes if n.id != node_id]
        # Clean up references
        for node in self.nodes:
            node.attacks = [a for a in node.attacks if a != node_id]
            node.supports = [s for s in node.supports if s != node_id]
        self.recalculate()

    def get_data(self) -> Dict[str, Any]:
        return {
            "colors": self.colors,
            "cats": self.cats,
            "globalScores": self.global_scores_config,
            "nodes": [n.to_dict() for n in self.nodes]
        }
